### ISSUE TITLE
✨ [Feature] Run with provided anatomical mask in 1.8

### ISSUE THREAD

**Issue by HechengJin0** [FCP-INDI/C-PAC#1485]
**Related problem**
copied an anatomical mask in /output dir to do brain extarctaction with provided mask. however, it requires json file as well. e.g. 
```
[!] No JSON found for file /outputs/output/cpac_cpac_default_monkey_skullstrip/sub-032125_ses-001/anat/sub-032125_ses-001_space-T1w_desc-brain_mask.nii.gz.
```

1. 

 if set {} in json file, 

```
Pulling outputs from /outputs/output.

Process Process-1:
Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 379, in run_workflow
    subject_id, sub_dict, c, p_name, num_ants_cores
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 1003, in build_workflow
    wf, rpool = initiate_rpool(wf, cfg, sub_dict)
  File "/code/CPAC/pipeline/engine.py", line 1826, in initiate_rpool
    rpool = ingress_output_dir(cfg, rpool, unique_id, creds_path)
  File "/code/CPAC/pipeline/engine.py", line 1535, in ingress_output_dir
    pipe_idx, node_name, inject=True)
  File "/code/CPAC/pipeline/engine.py", line 205, in set_data
    res, new_pipe_idx = self.generate_prov_string(new_prov_list)
  File "/code/CPAC/pipeline/engine.py", line 321, in generate_prov_string
    last_entry = get_last_prov_entry(prov)
  File "/code/CPAC/utils/utils.py", line 25, in get_last_prov_entry
    while not isinstance(prov[-1], str):
IndexError: list index out of range
```

2. 

 if blank json file

```
Pulling outputs from /outputs/output.

Process Process-1:
Traceback (most recent call last):
  File "/code/CPAC/utils/utils.py", line 121, in read_json
    json_dct = json.load(f)
  File "/usr/local/miniconda/lib/python3.7/json/__init__.py", line 296, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/local/miniconda/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/local/miniconda/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/miniconda/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 2 column 1 (char 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 379, in run_workflow
    subject_id, sub_dict, c, p_name, num_ants_cores
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 1003, in build_workflow
    wf, rpool = initiate_rpool(wf, cfg, sub_dict)
  File "/code/CPAC/pipeline/engine.py", line 1826, in initiate_rpool
    rpool = ingress_output_dir(cfg, rpool, unique_id, creds_path)
  File "/code/CPAC/pipeline/engine.py", line 1493, in ingress_output_dir
    json_info = read_json(jsonpath)
  File "/code/CPAC/utils/utils.py", line 123, in read_json
    raise Exception(f'\n\n{err}\n\nJSON file: {json_file}\n')
Exception:

Expecting value: line 2 column 1 (char 1)

JSON file: /outputs/output/cpac_cpac_default_monkey_skullstrip/sub-032125_ses-001/anat/sub-032125_ses-001_space-T1w_desc-brain_mask.json
```

3. 

 if no json file

```
Pulling outputs from /outputs/output.

Process Process-1:
Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 379, in run_workflow
    subject_id, sub_dict, c, p_name, num_ants_cores
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 1003, in build_workflow
    wf, rpool = initiate_rpool(wf, cfg, sub_dict)
  File "/code/CPAC/pipeline/engine.py", line 1826, in initiate_rpool
    rpool = ingress_output_dir(cfg, rpool, unique_id, creds_path)
  File "/code/CPAC/pipeline/engine.py", line 1490, in ingress_output_dir
    raise Exception('\n\n[!] No JSON found for file '
Exception: 

[!] No JSON found for file /outputs/output/cpac_cpac_default_monkey_skullstrip/sub-032125_ses-001/anat/sub-032125_ses-001_space-T1w_desc-brain_mask.nii.gz.
```

4. 

set json file

```
{
   "CpacProvenance": [],
   "RawSources": [
      "T1w"
   ],
   "Sources": [
      "desc-preproc_T1w"
   ]
}

and 
{
   "CpacProvenance": [],
   "RawSources": [
      "T2w"
   ],
   "Sources": [
      "desc-preproc_T2w"
   ]
}

```

```
Pulling outputs from /outputs/output.

Process Process-1:
Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 379, in run_workflow
    subject_id, sub_dict, c, p_name, num_ants_cores
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 1003, in build_workflow
    wf, rpool = initiate_rpool(wf, cfg, sub_dict)
  File "/code/CPAC/pipeline/engine.py", line 1826, in initiate_rpool
    rpool = ingress_output_dir(cfg, rpool, unique_id, creds_path)
  File "/code/CPAC/pipeline/engine.py", line 1519, in ingress_output_dir
    pipe_idx = rpool.generate_prov_string(json_info['CpacProvenance'])
  File "/code/CPAC/pipeline/engine.py", line 321, in generate_prov_string
    last_entry = get_last_prov_entry(prov)
  File "/code/CPAC/utils/utils.py", line 25, in get_last_prov_entry
    while not isinstance(prov[-1], str):
IndexError: list index out of range
```

**Comment by HechengJin0** [comment]
6. 

 set json file
```
{
   "CpacProvenance": [
      "T1w:anat_ingress",
      "desc-preproc_T1w:anatomical_init",
      "space-T1w_desc-brain_mask:brain_extraction" 
   ],
   "RawSources": [
      "T1w"
   ],
   "Sources": [
      "desc-preproc_T1w"
   ]
}

{
   "CpacProvenance": [
      "T2w:anat_ingress",
      "desc-preproc_T2w:anatomical_init_T2",
      "space-T2w_desc-brain_mask:brain_extraction_T2"
   ],
   "RawSources": [
      "T2w"
   ],
   "Sources": [
      "desc-preproc_T2w"
   ]
}
```

```Pulling outputs from /outputs/output.

Connecting anatomical_init...

Connecting brain_extraction...

Connecting acpc_alignment_brain_with_mask...

Connecting brain_extraction...

Connecting anatomical_init_T2...

Connecting brain_extraction_T2...

Connecting acpc_alignment_T2_brain_with_mask...

Process Process-1:
Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 379, in run_workflow
    subject_id, sub_dict, c, p_name, num_ants_cores
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 1232, in build_workflow
    wf = connect_pipeline(wf, cfg, rpool, pipeline_blocks)
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 958, in connect_pipeline
    wf = nb.connect_block(wf, cfg, rpool)
  File "/code/CPAC/pipeline/engine.py", line 1231, in connect_block
    self.check_output(outputs, label, name)
  File "/code/CPAC/pipeline/engine.py", line 1102, in check_output
    raise Exception('\n[!] Output name in the block function does '
Exception: 
[!] Output name in the block function does not match the outputs list in Node Block acpc_alignment_T2_brain_with_mask
```

7. 

 set json file
```
{
   "CpacProvenance": [
      "T1w:anat_ingress",
      "desc-preproc_T1w:anatomical_init",
      "space-T1w_desc-brain_mask:brain_extraction" 
   ],
   "RawSources": [
      "T1w"
   ],
   "Sources": [
      "desc-preproc_T1w"
   ]
}

{
   "CpacProvenance": [
      "T2w:anat_ingress"
   ],
   "RawSources": [
      "T2w"
   ],
   "Sources": [
      "desc-preproc_T2w"
   ]
}
```
```
Process Process-1:
Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 379, in run_workflow
    subject_id, sub_dict, c, p_name, num_ants_cores
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 1232, in build_workflow
    wf = connect_pipeline(wf, cfg, rpool, pipeline_blocks)
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 958, in connect_pipeline
    wf = nb.connect_block(wf, cfg, rpool)
  File "/code/CPAC/pipeline/engine.py", line 1193, in connect_block
    for pipe_idx, strat_pool in rpool.get_strats(inputs).items():         # strat_pool is a ResourcePool like {'desc-preproc_T1w': { 'json': info, 'data': (node, out) }, 'desc-brain_mask': etc.}
  File "/code/CPAC/pipeline/engine.py", line 464, in get_strats
    yjson = copy.deepcopy(json_dct[ylabel])
KeyError: "When trying to connect node block 'brain_extraction_T2' to workflow 'cpac_sub-032125_ses-001' after node block 'anatomical_init_T2': space-T2w_desc-brain_mask"

Pulling outputs from /outputs/output.

Connecting anatomical_init...

Connecting brain_extraction...

Connecting acpc_alignment_brain_with_mask...

Connecting brain_extraction...

Connecting anatomical_init_T2...

Connecting brain_extraction_T2...
```

**Comment by HechengJin0** [comment]
#6 is correct.

**Comment by sgiavasis** [comment]
This will be in v1.8.1 - CPAC will create a JSON for data injected into the output directory without one, but will make it clear this is what happened inside the JSON.

**Comment by shnizzedy** [comment]
Fixed in #1518
